op-challenger/game/fault/trace/split/provider.go

```diff
@@ -11,39 +13,89 @@ import (
+// used for the bottom half of the [SplitTraceProvider].
+type CreateLowerProvider func(ctx context.Context, preClaim common.Hash, postClaim common.Hash) (types.TraceProvider, error)
+
+// AbsolutePrestateCommitmentFetcher is a function pointer to fetch the absolute prestate commitment
+type AbsolutePrestateCommitmentFetcher func(ctx context.Context) (hash common.Hash, err error)
+
+// AbsolutePrestateFetcher is a function pointer to fetch the absolute prestate
+type AbsolutePrestateFetcher func(ctx context.Context) (preimage []byte, err error)
+
 // SplitTraceProvider is a [types.TraceProvider] implementation that
 // routes requests to the correct internal trace provider based on the
 // depth of the requested trace.
 type SplitTraceProvider struct {
-	logger         log.Logger
-	topProvider    types.TraceProvider
-	bottomProvider types.TraceProvider
-	topDepth       uint64
+	logger                            log.Logger
+	upper                             types.TraceProvider
+	lower                             map[*big.Int]types.TraceProvider
+	lowerCreator                      CreateLowerProvider
+	absolutePrestateCommitmentFetcher AbsolutePrestateCommitmentFetcher
+	absolutePrestateFetcher           AbsolutePrestateFetcher
+	topDepth                          uint64
 }
 
 // NewTraceProvider creates a new [SplitTraceProvider] instance.
 // The [topDepth] parameter specifies the depth at which the internal
 // [types.TraceProvider] should be switched.
-func NewTraceProvider(logger log.Logger, topProvider types.TraceProvider, bottomProvider types.TraceProvider, topDepth uint64) *SplitTraceProvider {
+func NewTraceProvider(logger log.Logger, upper types.TraceProvider, creator CreateLowerProvider, topDepth uint64) *SplitTraceProvider {
 	return &SplitTraceProvider{
-		logger:         logger,
-		topProvider:    topProvider,
-		bottomProvider: bottomProvider,
-		topDepth:       topDepth,
+		logger:       logger,
+		upper:        upper,
+		lower:        make(map[*big.Int]types.TraceProvider),
+		lowerCreator: creator,
+		topDepth:     topDepth,
 	}
 }
 
-func (s *SplitTraceProvider) providerForDepth(depth uint64) (uint64, types.TraceProvider) {
-	if depth <= s.topDepth {
-		return 0, s.topProvider
+func (s *SplitTraceProvider) buildLowerProvider(ctx context.Context, pos types.Position) (types.TraceProvider, error) {
+	ancestorIndex := pos.IndexForParentDepth(int(s.topDepth))
+	provider, ok := s.lower[ancestorIndex]
+	if ok {
+		return provider, nil
+	}
+
+	// Fetch the preclaim and postclaim for the ancestor.
+	leftIndex := new(big.Int).Sub(ancestorIndex, big.NewInt(1))
+	rightIndex := new(big.Int).Add(ancestorIndex, big.NewInt(1))
+	prePos := types.NewPosition(int(s.topDepth), leftIndex)
+	postPos := types.NewPosition(int(s.topDepth), rightIndex)
+	preClaim, err := s.upper.Get(ctx, prePos)
+	if err != nil {
+		return nil, err
 	}
-	return s.topDepth, s.bottomProvider
+	postClaim, err := s.upper.Get(ctx, postPos)
```

Comment:
This is nice and simple but unfortunately doesn't work.  The pre and post claims must come from the claims made in the game, not from our trace provider, otherwise they may not match what is used on chain.  One or either of the pre and post-claim values may have been provided by our opponent and we're trying to disprove it.  The input to `op-program` should be the output root they provide and the cannon trace we claim to have must then have the invalid status.

We need to walk back up the claim ancestors until we find the ones that are from the trace index we need for pre and post claims.